### PR TITLE
Support COM objects with dynamic keyword

### PR DESCRIPTION
**Pull request by elinor-fung**
This is a port of the support (which is in Framework and the IronLanguages DLR) for using the dynamic keyword with COM objects.

The actual support is in `Microsoft.CSharp`. The tests are under coreclr, since that is set up to test against a native COM server. I did a couple passes to make the ported sources fit in the runtime repo (I'm sure there's more that could be done).

The vast majority of this is a simple copy of the previous support. The commits in this PR are:
- Straight copy of all the sources (first commit [21a9e9a](https://github.com/dotnet/runtime/pull/33060/commits/21a9e9a754e2453bb90a1e8a362e04460b219a28))
- Changes (next 15 commits) to get it building and working
- New tests
- Additional clean-up/consolidation to make it fit a bit better in this repo.

cc @jaredpar @cston 

Resolves #12587

### CODE REVIEW

File: src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs [hunk 1]
```diff
@@ -2,6 +2,8 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#nullable enable
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs]: Why is this necessary?  Is this file built into some project other than Corelib and that hasn't been annotated yet for nullable reference types? (Same question for the other occurrences in other files.)

File: src/libraries/Common/src/System/Runtime/InteropServices/Variant.cs [hunk 2]
```diff
@@ -170,9 +174,11 @@ public unsafe void CopyFromIndirect(object value)
                     *(uint*)this._typeUnion._unionTypes._byref = (uint)value;
                     break;
 
+#pragma warning disable 618
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Runtime/InteropServices/Variant.cs]: What required this be added?  IIRC, 0618 is for using an obsoleted member.

File: src/libraries/Microsoft.CSharp/src/ILLinkTrim.xml [hunk 3]
```diff
@@ -4,5 +4,7 @@
     <type fullname="Microsoft.CSharp.RuntimeBinder.DynamicMetaObjectProviderDebugView"/>
     <!-- Required by the Expression Evaluator -->
     <type fullname="Microsoft.CSharp.RuntimeBinder.DynamicMetaObjectProviderDebugView/DynamicDebugViewEmptyException"/>
+    <!-- Required for COM event dispatch -->
+    <type fullname="System.Runtime.InteropServices.ComEventsSink"/>
```
Comment (stephentoub) [inline src/libraries/Microsoft.CSharp/src/ILLinkTrim.xml]: This prevents ComEventsSink in Microsoft.CSharp.dll from being trimmed out.  Is it being accessed by reflection somewhere and that's why this is needed?

File: src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs (hunk 1 above)
Comment (elinor-fung) [reply src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs]: `Microsoft.CSharp` has not been

File: src/libraries/Common/src/System/Runtime/InteropServices/Variant.cs (hunk 2 above)
Comment (elinor-fung) [reply src/libraries/Common/src/System/Runtime/InteropServices/Variant.cs]: `ErrorWrapper` is obsolete. I'll add a comment. (Same with the other `*Wrapper` classes for COM).

File: src/libraries/Microsoft.CSharp/src/ILLinkTrim.xml (hunk 3 above)
Comment (elinor-fung) [reply src/libraries/Microsoft.CSharp/src/ILLinkTrim.xml]: Parts of it will be accessed by whatever COM server is firing the events for which a handler has been added. Specifically, it is the `ComEventsSink`'s implementation of `IDispatch` that is not explicitly used (directly or via reflection), but is necessary for handling COM events.

Is there a better way to prevent that from being trimmed out? I know there is `PreserveDependency`, but since we are not actually the ones that call into it, we don't actually have another method that is using it.

Comment (stephentoub) [reply src/libraries/Microsoft.CSharp/src/ILLinkTrim.xml]: Thanks for the explanation.  I think what you have is the best we can currently do.  That said, are you actually seeing the interface implementation get removed?  I didn't think the linker would do that if the interface is used anywhere, but my knowledge may be out of date.

File: src/libraries/Common/src/System/Runtime/InteropServices/Variant.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/Common/src/System/Runtime/InteropServices/Variant.cs]: > ErrorWrapper is obsolete

Thanks.  My question was really, why wasn't this suppression necessary before?

File: src/libraries/Microsoft.CSharp/src/ILLinkTrim.xml (hunk 3 above)
Comment (elinor-fung) [reply src/libraries/Microsoft.CSharp/src/ILLinkTrim.xml]: Yeah, I hit this because the implementation of `IDispatch.Invoke` was getting removed.

File: src/libraries/Common/src/System/Runtime/InteropServices/Variant.cs (hunk 2 above)
Comment (elinor-fung) [reply src/libraries/Common/src/System/Runtime/InteropServices/Variant.cs]: Ahh, it looks like this is because those wrapper classes are marked obsolete in the ref assembly for System.Runtime.InteropServices, but not the implementation. So when this file was only used by CoreLib and not going through the ref assembly, it was not considered obsolete.

Comment (stephentoub) [reply src/libraries/Common/src/System/Runtime/InteropServices/Variant.cs]: I see. Ok, thanks.

File: src/libraries/Microsoft.CSharp/src/ILLinkTrim.xml (hunk 3 above)
Comment (jkotas) [reply src/libraries/Microsoft.CSharp/src/ILLinkTrim.xml]: https://github.com/mono/linker/issues/936 ?

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs [hunk 4]
```diff
@@ -436,7 +442,20 @@ internal static object TryGetMemberValue(object obj, string name, bool ignoreExc
 
                 return result;
             }
-
+#if ENABLECOMBINDER
+            else if (obj != null && ComObjectType.IsAssignableFrom(obj.GetType()))
+            {
+                var comExclusionList = new string[] { "MailEnvelope" }; //add any com names to be excluded from dynamic view here
```
Comment (VSadov) [inline src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs]: >MailEnvelope [](start = 55, length = 12)

I wonder what is the history behind excluding `MailEnvelope` from the debug view and whether the exclusion is still relevant.

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+        }
+
+        public static DispCallable CreateDispCallable(IDispatchComObject dispatch, ComMethodDesc method)
+        {
+            return new DispCallable(dispatch, method.Name, method.DispId);
+        }
+
+        internal static MethodInfo GetGetIDispatchForObjectMethod()
+        {
+            // GetIDispatchForObject always throws a PNSE in .NET Core, so we work around it by using GetComInterfaceForObject with our IDispatch type.
```
Comment (jkotas) [inline src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs]: Would it make sense to change `GetIDispatchForObject` to forward to `GetComInterfaceForObject` in .NET Core?

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs [hunk 6]
```diff
@@ -0,0 +1,578 @@
...
+            if (obj is IDispatch)
+            {
+                variant.AsDispatch = obj;
+                return;
+            }
+
+            Marshal.GetNativeVariantForObject(obj, ConvertVariantByrefToPtr(ref variant));
+        }
+
+        [Obsolete("do not use this method", true)]
```
Comment (jkotas) [inline src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs]: These Obsolete attributes do not look useful. Are these methods accessed by reflection only? If yes, it should be enough to just add a comment. If not, delete them.

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+                if (s_dynamicModule != null)
+                {
+                    return s_dynamicModule;
+                }
+                lock (s_lock)
+                {
+                    if (s_dynamicModule == null)
+                    {
+                        var attributes = new CustomAttributeBuilder[] {
+                            new CustomAttributeBuilder(typeof(UnverifiableCodeAttribute).GetConstructor(Array.Empty<Type>()), Array.Empty<object>())
```
Comment (jkotas) [inline src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs]: UnverifiableCodeAttribute does nothing in .NET Core. No reason to emit it.

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs (hunk 6 above)
Comment (elinor-fung) [reply src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs]: Yep - accessed via reflection (seems like adding an Obsolete attribute like this was common in framework for parts of System.Dynamic). Will remove and switch to a comment.

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComBinderHelpers.cs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            return mo.Expression is ParameterExpression pe && pe.IsByRef;
+        }
+
+        internal static bool IsStrongBoxArg(DynamicMetaObject o)
+        {
+            Type t = o.LimitType;
+            return t.IsGenericType && t.GetGenericTypeDefinition() == typeof(StrongBox<>);
+        }
+
+        // this helper prepares arguments for COM binding by transforming ByVal StongBox arguments
```
Comment (VSadov) [inline src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComBinderHelpers.cs]: >StongBox [](start = 80, length = 8)

StrongBox   (typo)

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComEventSinksContainer.cs [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        private ComEventSinksContainer()
+        {
+        }
+
+        private static readonly object s_comObjectEventSinksKey = new object();
+
+        public static ComEventSinksContainer FromRuntimeCallableWrapper(object rcw, bool createIfNotFound)
+        {
+            // !!! Marshal.Get/SetComObjectData has a LinkDemand for UnmanagedCode which will turn into
```
Comment (VSadov) [inline src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComEventSinksContainer.cs]: The comment about LinkDemand and SecurityCritical can be removed now :-)

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs (hunk 4 above)
Comment (elinor-fung) [reply src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/DynamicDebuggerProxy.cs]: That's all there is in Framework as far as comments go. Looks like it has been there since the sources were added / copied in 2010 (not sure how to go back further).

I tried removing the exclusion. And was able to see the `MailEnvelope` property for an Excel `Worksheet` and a Word `Document` just fine.

File: src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs [hunk 10]
```diff
@@ -220,6 +225,40 @@ public void RemoveDelegate(Delegate d)
...
+                                toRemove.Add(i);
+                            }
+                        }
+                    }
+                }
+
+                foreach (int idx in toRemove)
+                {
+                    _delegateWrappers.RemoveAt(idx);
+                }
```
Comment (cston) [inline src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs]: Do we need to remove higher indices before lower?

File: src/libraries/Common/src/System/Runtime/InteropServices/ComEventsSink.cs [hunk 11]
```diff
@@ -10,7 +12,7 @@ namespace System.Runtime.InteropServices
     /// Part of ComEventHelpers APIs which allow binding
     /// managed delegates to COM's connection point based events.
     /// </summary>
-    internal class ComEventsSink : IDispatch, ICustomQueryInterface
+    internal partial class ComEventsSink : IDispatch, ICustomQueryInterface
```
Comment (VSadov) [inline src/libraries/Common/src/System/Runtime/InteropServices/ComEventsSink.cs]: >IDispatch, ICustomQueryInterface [](start = 43, length = 32)

Nice! In olden time this had to use `IReflect/RealProxy` stuff to pretend it implements interfaces.

File: src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs (hunk 10 above)
Comment (elinor-fung) [reply src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs]: Ack - yes.

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComObject.cs [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+        /// <summary>
+        /// Gets a <see cref="ComObject"/> that wraps the runtime-callable-wrapper, or creates one if none currently exists.
+        /// </summary>
+        /// <returns></returns>
+        public static ComObject ObjectToComObject(object rcw)
+        {
+            Debug.Assert(ComBinder.IsComObject(rcw));
+
+            // Marshal.Get/SetComObjectData has a LinkDemand for UnmanagedCode which will turn into
```
Comment (VSadov) [inline src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComObject.cs]: >// Marshal.Get/SetComObjectData has a LinkDeman [](start = 12, length = 47)

Another mention of LinkDemand

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/Helpers.cs [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+                return Expression.Block(expression, Expression.Empty());
+            }
+
+            return Expression.Convert(expression, type);
+        }
+    }
+
+    internal static class Requires
+    {
+        internal static void NotNull(object value, string paramName)
```
Comment (VSadov) [inline src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/Helpers.cs]: >NotNull [](start = 29, length = 7)

This made sense when this was public API. As long as the COM binder is internal, this could be `[Conditional("DEBUG")]`

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/Helpers.cs [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        {
+            if (value == null)
+            {
+                throw new ArgumentNullException(paramName);
+            }
+        }
+
+        internal static void Condition(bool precondition, string paramName)
```
Comment (VSadov) [inline src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/Helpers.cs]: This made sense when this was public API. As long as the COM binder is internal, this could be `[Conditional("DEBUG")]`

File: src/coreclr/tests/src/Interop/COM/Dynamic/Server/DispatchImpl.h [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    HRESULT DoGetIDsOfNames(LPOLESTR* rgszNames, UINT cNames, DISPID* rgDispId);
+    HRESULT DoInvoke(DISPID dispIdMember, WORD wFlags, DISPPARAMS* pDispParams, VARIANT* pVarResult, EXCEPINFO* pExcepInfo, UINT* puArgErr);
+
+private:
+    ITypeLib *_typeLib;
+    ITypeInfo *_typeInfo;
+    void *_instance;
+};
+
+// Marco to use for defining ref counting impls
```
Comment (cston) [inline src/coreclr/tests/src/Interop/COM/Dynamic/Server/DispatchImpl.h]: >Marco [](start = 3, length = 5)

Typo

File: src/coreclr/tests/src/Interop/COM/Dynamic/Program.cs [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    using System;
+    using TestLibrary;
+
+    /*
+       Param array : ParamArrayArgs
+     */
+
+    class Program
+    {
+        static int Main(string[] doNotUse)
```
Comment (cston) [inline src/coreclr/tests/src/Interop/COM/Dynamic/Program.cs]: >Main [](start = 19, length = 4)

Are these tests run in CI builds?

File: src/coreclr/tests/src/Interop/COM/Dynamic/Program.cs [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+// See the LICENSE file in the project root for more information.
+
+namespace Dynamic
+{
+    using System;
+    using TestLibrary;
+
+    /*
+       Param array : ParamArrayArgs
+     */
```
[the remark this answers is not in the document]
Comment (elinor-fung) [reply src/coreclr/tests/src/Interop/COM/Dynamic/Program.cs]: Bah - left over from notes to myself about tests to add. Will remove.

File: src/coreclr/tests/src/Interop/COM/Dynamic/Program.cs (hunk 16 above)
Comment (elinor-fung) [reply src/coreclr/tests/src/Interop/COM/Dynamic/Program.cs]: Yes (as part of the CoreCLR tests)

**Comment by VSadov** [comment]
I left some comments, but there is nothing that is blocking. The change looks nice. 
- I like that the remoting/proxy pieces are no longer needed.
- It is nice to see that most IL-generated helpers (like ConvertByrefToPtr  flavors) are gone. Maybe one day we can get rid of the helpers that wrap `CALLI` as well.

LGTM

**Review by VSadov — APPROVED** [review]

**Review by cston — APPROVED** [review]

File: src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs (hunk 5 above)
Comment (elinor-fung) [reply src/libraries/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/ComInterop/ComRuntimeHelpers.cs]: I think that makes sense. It looks like we also have `MarshalNative::GetIDispatchForObjectNative` (currently unused, I think).

I filed #33217 for it.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by jaredpar** [comment]
@VSadov 

> It is nice to see that most IL-generated helpers (like ConvertByrefToPtr flavors) are gone. Maybe one day we can get rid of the helpers that wrap CALLI as well.

Is this something that function pointers will help out with?  @333fred

**Comment by VSadov** [comment]
@jaredpar - Yes, I think the whole point of function pointers is to make CALLI expressible in C# and handle scenarios like this.
